plugins/package-managers/bazel/src/main/kotlin/Bazel.kt

```diff
@@ -0,0 +1,221 @@
+
+    override fun resolveDependencies(definitionFile: File, labels: Map<String, String>): List<ProjectAnalyzerResult> {
+        val projectDir = definitionFile.parentFile
+        val lockFile = projectDir.resolve(LOCKFILE_NAME)
+        val projectVcs = processProjectVcs(projectDir)
+        val moduleMetadata = Parser(definitionFile.readText()).parse()
+        val depDirectives = moduleMetadata.dependencies.associateBy { "${it.name}@${it.version}" }
+        val scopes = getDependencyGraph(projectDir, depDirectives)
+
+        // If no lock file is present, getDependencyGraph() runs "bazel mod graph", which creates a MODULE.bazel.lock
+        // file as a side effect. That file contains the URL of the Bazel module registry that was used for dependency
+        // resolution.
+        val lockFileData = parseLockfile(lockFile)
+
+        val bmrClient = BazelModuleRegistryClient.create(lockFileData.registryUrl())
+        val packages = scopes.flatMap { getPackages(it.dependencies, bmrClient) }.filterNotNull()
+
+        return listOf(
+            ProjectAnalyzerResult(
+                project = Project(
+                    id = Identifier(
+                        type = managerName,
+                        name = moduleMetadata.module?.name ?: "root",
+                        version = moduleMetadata.module?.version.orEmpty(),
+                        namespace = ""
+                    ),
+                    definitionFilePath = VersionControlSystem.getPathInfo(definitionFile).path,
+                    authors = emptySet(),
+                    declaredLicenses = emptySet(),
+                    vcs = projectVcs,
+                    vcsProcessed = projectVcs,
+                    homepageUrl = "",
+                    scopeDependencies = scopes
+                ),
+                packages = packages.toSet()
+            )
+        )
+    }
+
+    private fun getPackages(
+        graph: Set<PackageReference>,
+        bmrClient: BazelModuleRegistryClient,
+        seen: MutableSet<Identifier> = mutableSetOf()
+    ): Set<Package?> {
+        val result = mutableSetOf<Package?>()
+
+        graph.forEach { ref ->
+            if (ref.id !in seen) {
+                seen.add(ref.id)
+
+                if (ref.id.name != "<root>") {
+                    result.add(packageFromPackageReference(ref, bmrClient))
+                }
+                result.addAll(getPackages(ref.dependencies, bmrClient, seen))
+            }
+        }
+
+        return result
+    }
+
+    private fun packageFromPackageReference(ref: PackageReference, bmrClient: BazelModuleRegistryClient): Package? {
+        val metadata: ModuleMetadata
+        try {
+            metadata = runBlocking { bmrClient.getModuleMetadata(ref.id.name) }
+        } catch (e: HttpException) {
+            logger.warn { "Failed to fetch metadata for Bazel module ${ref.id.name}: $e" }
+            return null
```

Comment:
Should we return a `Package` with just the `id` instead of a `null` one, to not have dangling references in the result? (Same for the `null` return below).